src/image.rs

```diff
@@ -412,6 +412,11 @@ pub trait ImageDecoder<'a>: Sized {
     fn into_reader(self) -> ImageResult<Self::Reader>;
 
     /// Returns the total number of bytes in the decoded image.
+    ///
+    /// This is the size of the buffer that must be passed to `read_image` or
```

Comment:
Hm, that seems backwards. Note that partial reads need to be supported in some form anyways since `Self::Reader` requires it from being an `io::Read`. Allowing reading a part of the image without reading it completely sounds incredibly useful.

Reply:
The _read_rect_ / _read_rect_with_progress_ functions on ImageDecoderExt allow partial reads. If decoders can't provide efficient random access but can still incrementally decode then they can provide a non-trivial reader implementation (i.e. one that doesn't just read from a Cursor<Vec<u8>>). If they can do neither then _read_image_ is intended as their primary interface and everything else will be implemented in terms of it.

Reply:
> The read_rect / read_rect_with_progress functions on ImageDecoderExt allow partial reads.

I think neither `ImageDecoderExt` nor `Self::Reader: io::Read` are very good candidates for an optimal partial read interface. Let me explain:

The `Reader` implementation can cover *some* of this but not all. With `io::Read` it offers a very standard but also inflexible interface. Part of that is the error type not being under our control, no conversion back and hence losing the access to the metadata, the implictness of the single possible order of pixels within the stream, ...

Offering a standard `io::Read` interface, the `Reader` can't optimize for discarding parts since every pixel *will* be loaded into the buffer at some point. It can't speed up with a seek on the underlying reader, and this means every single pixel will be loaded, decoded, inspected. Adding `Seek` to such an implementation sounds possible but requiring it in the trait on the other hand would be a complication especially since then `Seek` will also be required on the underlying stream where it currently is not for some decoders. This together with the fact that we don't even want a generic `Seek`, e.g. don't want to go backwards, greatly outweighs the potential benefits for me.

In their most prominent implementation, `ImageDecoderExt` also doesn't provide a big benefit or none at all. The only actual implementation available reads scanlines at a time (iirc) which still requires a *huge* buffer and this also requires, as explained above, *the exact same throughput* as reading the complete image.

And also importantly, `ImageDecoderExt` is not always available and especially not when one wants a generic interface over all decoders. The typical `*Ext` naming is for traits that purely extend an interface and are implementation for all `T: BaseTrait`, so this is also a confusing naming.

Reply:
I don't think there are any perfect solutions to incremental image decoding. I think the current options we have represent good points in the trade-off space, but I'd be open to suggestions on alternatives. Really, a lot of the issue right now is that the current slate of decoders don't actually have much in the way of incremental decoding support. There's no reason why not, other than no one taking the time to implement them. Just about ever decoder's Reader<R> could easily also implement Seek when R did, and almost every format can use Seek to make a rather good read_rect implementation.

> In their most prominent implementation, ImageDecoderExt also doesn't provide a big benefit or none at all. The only actual implementation available reads scanlines at a time (iirc) which still requires a huge buffer and this also requires, as explained above, the exact same throughput as reading the complete image.

The only "real" implementation of ImageDecoderExt is for DXT. Here it actually does seek to the first row of the target rect, but then reads/decodes 4 entire rows at a time (a "scanline") and splices the needed ranges into the target buffer. Depending on shape of the rect, it may or may not be faster to use seeking to only read the parts of rows that are needed.

> The typical *Ext naming is for traits that purely extend an interface and are implementation for all T: BaseTrait, so this is also a confusing naming.

This is a very good point. Any ideas for alternate names? Perhaps `ImageDecoderSeekable` or something?

Reply:
>  The only "real" implementation of ImageDecoderExt is for DXT. Here it actually does seek to the first row of the target rect, but then reads/decodes 4 entire rows at a time (a "scanline") and splices the needed ranges into the target buffer. Depending on shape of the rect, it may or may not be faster to use seeking to only read the parts of rows that are needed.

That seems like an important point in itself: Image formats are explicitely not built for a linear stream of values, and thus the locality of values will not be linear at all but sometimes/often block based. Having the complete target buffer available is sometimes a real advantage that an `io::Read`-based implementation can never enjoy!

Reply:
Just because it doesn't seem as clear on a second read: This means that I think the requirement of passing in the whole decoding buffer is the *correct* choice since it is a very minimal interface that guarantees that an output region is available no matter the ordering of pixels within the encoded format.